File: lib/model/sensor_data.dart

/// Class representing the single raw data captured from sensors
///
/// This class contains the data obtained from all the sensors
/// during a single event, so a list of [SensorData] will represent
/// all the sensors data collected during a test.
/// The data collected are:
/// - timestamp
/// - accuracy
/// - accelerometer x,y,z
/// - gyroscope x,y,z
class SensorData {
  final int timestamp;
  final int accuracy;
  final double accelerometerX;
  final double accelerometerY;
  final double accelerometerZ;
  final double gyroscopeX;
  final double gyroscopeY;
  final double gyroscopeZ;

  SensorData(
    this.timestamp,
    this.accuracy,
    this.accelerometerX,
    this.accelerometerY,
    this.accelerometerZ,
    this.gyroscopeX,
    this.gyroscopeY,
    this.gyroscopeZ,
  );

  @override
  String toString() => "SensorData(timestamp=$timestamp, "
    "accuracy=$accuracy, "
    "accelerometerX=$accelerometerX, "
    "accelerometerY=$accelerometerY, "
    "accelerometerZ=$accelerometerZ, "
    "gyroscopeX=$gyroscopeX, "
    "gyroscopeY=$gyroscopeY, "
    "gyroscopeZ=$gyroscopeZ)";

  @override
  bool operator ==(other) =>
    other is SensorData &&
      other.timestamp == this.timestamp &&
      other.accuracy == this.accuracy &&
      other.accelerometerX == this.accelerometerX &&
      other.accelerometerY == this.accelerometerY &&
      other.accelerometerZ == this.accelerometerZ &&
      other.gyroscopeX == this.gyroscopeX &&
      other.gyroscopeY == this.gyroscopeY &&
      other.gyroscopeZ == this.gyroscopeZ;

  @override
  int get hashCode {
    final int prime = 31;
    int result = 1;
    result = prime * result + timestamp.hashCode;
    result = prime * result + accuracy.hashCode;
    result = prime * result + accelerometerX.hashCode;
    result = prime * result + accelerometerY.hashCode;
    result = prime * result + accelerometerZ.hashCode;
    result = prime * result + gyroscopeX.hashCode;
    result = prime * result + gyroscopeY.hashCode;
    result = prime * result + gyroscopeZ.hashCode;
    return result;
  }
}

File: android/app/src/main/kotlin/srl/digit/balance/model/SensorData.kt
package srl.digit.balance.model

/**
 * Class representing the data collected by the sensors
 *
 * This class stores one value retrieved by the accelerometer
 * or gyroscope sensors
 *
 * @param timestamp timestamp of the data
 * @param accuracy the accuracy of the sensors
 * @param x values along x axes
 * @param y values along x axes
 * @param z values along x axes
 *
 * @see srl.digit.balance.sensor.SensorMonitor
 * @see srl.digit.balance.sensor.SensorSharedValues
 * @see srl.digit.balance.sensor.SensorListener
 * @author Lorenzo Calisti on 09/03/2020
 */
data class SensorData(
	val timestamp: Long = 0L,
	val accuracy: Int = 0,
	val x: Float? = null,
	val y: Float? = null,
	val z: Float? = null
) {

	companion object {
		/**
		 * Merge two [SensorData] in a list
		 *
		 * Returns a list given by the combination of two [SensorData];
		 * the list will be then passed to Flutter in the [io.flutter.plugin.common.EventChannel].
		 * The list contains the data in the following order:
		 * 	- timestamp
		 * 	- accuracy
		 * 	- accelerometerX
		 * 	- accelerometerY
		 * 	- accelerometerZ
		 * 	- gyroscopeX
		 * 	- gyroscopeY
		 * 	- gyroscopeZ
		 *
		 * @param acc accelerometer [SensorData]
		 * @param gyr gyroscope [SensorData]
		 * @return List with the combination of the SensorData
		 */
		fun mergeSensorData(acc: SensorData?, gyr: SensorData?): List<Number?>? =
			when {
				// Merge the two data into one
				(acc != null && gyr != null) ->
					listOf(
						maxOf(acc.timestamp, gyr.timestamp),
						maxOf(acc.accuracy, gyr.accuracy),
						acc.x,
						acc.y,
						acc.z,
						gyr.x,
						gyr.y,
						gyr.z
					)
				// The gyroscope data are null
				acc != null -> listOf(acc.timestamp, acc.accuracy, acc.x, acc.y, acc.z, null, null, null)
				// The accelerometer data are null
				gyr != null -> listOf(gyr.timestamp, gyr.accuracy, null, null, null, gyr.x, gyr.y, gyr.z)
				// Both data are null, we cannot merge
				else -> null
			}
	}
}
